tools/autograd/gen_variable_type.py

```diff
@@ -639,21 +649,23 @@ def save_variables(
                 stmts.append('}')
         return stmts
 
-    def emit_dispatch_call(f: NativeFunction, input_base: str, unpacked_args: Sequence[str]) -> str:
+    def emit_dispatch_call(f: NativeFunction, input_base: str, unpacked_args: Sequence[str], *, mask_out_autograd: bool) -> str:
         """ Dispatch call via function in a namespace or method on Tensor."""
-        if Variant.function in f.variants:
-            call = CALL_DISPATCH_VIA_NAMESPACE.substitute(
-                api_name=cpp.name(
-                    f.func,
-                    faithful_name_for_out_overloads=True,
-                ),
-                unpacked_args=unpacked_args)
+        dispatcher_sig = DispatcherSignature.from_schema(f.func)
+        dispatcher_exprs = dispatcher_sig.exprs()
+        # code-generated autograd kernels plumb and recompute dispatch keys directly through the kernel for performance.
+        # See Note [Plumbing Keys Through The Dispatcher] for details.
+        if mask_out_autograd:
+            dispatch_key_set = 'ks & c10::after_autograd_keyset'
         else:
-            call = CALL_DISPATCH_VIA_METHOD.substitute(
-                api_name=cpp.name(f.func),
-                var=input_base,
-                unpacked_method_args=unpacked_args[1:])
-        return call
+            # view functions need to be able to re-run autograd
```

Comment:
What's going on here?

Reply:
Oh, this is invoked when `mask_out_autograd=False`. It's not clear to me that the formulation here makes sense. cc @ailzhang 

What is going on with view replay is that in some situations (specifically, XLA and complex dtype coercion), we can't just look at strides to reconstruct an appropriate view metadata when the underlying metadata of the base tensor changes. So we "replay" the view operations and then use the resulting metadata in that case to populate the normal view operation.

In the old implementation, this replay happened as normal function calls in whatever the ambient context at the time. With this change, you are pinning the view replay to whatever the context was *at the time the view replay function was recorded.* It is hard to think of concrete examples today which will break due to this, but it seems to me that morally the ambient context is more correct (e.g., what if the dispatch keys of some of the input tensors have mutated in the meantime? A prerecorded dispatch key set would just be wrong).

Reply:
From what I gathered, autograd kernels generate two lambdas: a redispatching lambda that removes autograd keys (via TLS) and is always executed, and an optional "view" lambda that redispatching **without** removing autograd keys, and is only executed later in autograd in for view functions. This is a little out-dated, but a built example for `view_as_complex` is [here](https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/autograd/generated/VariableType_0.cpp.html#9079).

My goal was to replicate that behavior using `Dispatcher::redispatch()` instead of `Dispatcher::call()`- since this lambda doesn't use the `nonVariableTypeMode` guard, I did not mask out autograd keys. (I actually ran into an awful bug originally when I did do the masking, it causes the backwards function to break for some view functions).

Looking at it again, I can imagine scenarios where this new behavior could in theory diverge from the existing behavior. If there are any dispatch keys with a higher priority than Autograd, that have an associated kernel that does **not** exclude its key using TLS, then:
- The current version of master would cause the autograd kernel's redispatch to hit that kernel a second time, using `Dispatcher::call()`
- The version in my PR would not redispatch to that kernel a second time, because `redispatch(ks)` and `redispatch(ks & c10::after_autograd_keyset)` have both masked out all kernels with higher priority than autograd.

That's technically BC breaking. I'm not sure how big of a deal that is, given:
(a) It's not causing any tests failures in pytorch core or fbcode
(b) The behavior you'd have to exercise to get there seems wrong to me: you'd be "going backwards" in your dispatch chain, from a kernel of lower priority to one of higher priority. Part of the contract (?) of using the dispatcher is that you shouldn't go backwards in the chain for the same op, at the risk of infinite loops (although it's possible to do so and avoid the infinite loops, e.g. with this view lambda).

Reply:
Oh, given what you said about the pinning the replay to the context when it was created being "morally less correct", I can also pretty easily just convert this back to the original code with the following (minor) downsides:
- autograd codegen is marginally more complex, since it now has to know how to use both dispatcher API calls
- the replay view lambda will be marginally more expensive, which I don't expect to be an issue since it's doesn't look like a hot path

Reply:
> autograd codegen is marginally more complex, since it now has to know how to use both dispatcher API calls

Yes. Basically you don't get to delete the old code. It's not too complicated though.

> the replay view lambda will be marginally more expensive, which I don't expect to be an issue since it's doesn't look like a hot path

Agreed